<launch>

  <arg name="x" default="0.0"/>
  <arg name="y" default="0.0"/>
  <arg name="z" default="0.0"/>
  <arg name="robot_name" default="turtlebot" />

  <!-- Gazebo model spawner for turtlebot -->
  <node name="spawn_$(arg robot_name)" pkg="gazebo_ros" type="spawn_model"
        args="-param /robot_description                                                                                                               
              -urdf                                                                                                                                   
              -unpause                                                                                                                                
              -x $(arg x)                                                                                                                             
              -y $(arg y)                                                                                                                             
              -z $(arg z)                                                                                                                             
              -model $(arg robot_name)                                                                                                                
              " respawn="false" output="screen">
  </node>


  <node pkg="nodelet" type="nodelet" name="$(arg robot_name)_nodelet_manager" args="manager"/>

  <node pkg="nodelet" type="nodelet" name="cmd_vel_mux"
        args="load yocs_cmd_vel_mux/CmdVelMuxNodelet $(arg robot_name)_nodelet_manager">
    <param name="yaml_cfg_file" value="$(find turtlebot_bringup)/param/mux.yaml" />
    <remap from="cmd_vel_mux/output" to="commands/velocity"/>
  </node>

  <!-- bumper2pc.launch.xml -->
  <node pkg="nodelet" type="nodelet" name="bumper2pointcloud" args="load kobuki_bumper2pc/Bumper2PcNodelet $(arg robot_name)_nodelet_manager">
    <param name="pointcloud_radius" value="0.24"/>
    <remap from="bumper2pointcloud/pointcloud"   to="sensors/bumper_pointcloud"/>
    <remap from="bumper2pointcloud/core_sensors" to="sensors/core"/>
  </node>

<!-- robot_state_publisher  -->
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="30.0" />
    <param name="tf_prefix" type="string" value="$(arg robot_name)" />
  </node>

<!-- fake lazer -->
  <node pkg="nodelet" type="nodelet" name="$(arg robot_name)_laserscan_nodelet_manager" args="manager"/>
  <node pkg="nodelet" type="nodelet" name="depthimage_to_laserscan"
        args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet $(arg robot_name)_laserscan_nodelet_manager">
    <param name="scan_height" value="10"/>
    <param name="output_frame_id" value="$(arg robot_name)/camera_depth_frame"/>
    <param name="range_min" value="0.45"/>
    <remap from="image" to="camera/depth/image_raw"/>
    <remap from="scan" to="scan"/>
  </node>

<!-- Velocity smoother -->
  <node pkg="nodelet" type="nodelet" name="$(arg robot_name)_navigation_velocity_smoother" args="load yocs_velocity_smoother/VelocitySmootherNodelet $(arg robot_name)_nodelet_manager">
    <rosparam file="$(find turtlebot_bringup)/param/defaults/smoother.yaml" command="load"/>
    <remap from="$(arg robot_name)_navigation_velocity_smoother/smooth_cmd_vel" to="cmd_vel_mux/input/navi"/>
    <remap from="$(arg robot_name)_navigation_velocity_smoother/odometry" to="odom"/>
    <remap from="$(arg robot_name)_navigation_velocity_smoother/robot_cmd_vel" to="commands/velocity"/>
  </node>

<!-- Safety controller-->
  <node pkg="nodelet" type="nodelet" name="$(arg robot_name)_kobuki_safety_controller" args="load kobuki_safety_controller/SafetyControllerNodelet $(arg robot_name)_nodelet_manager">
    <remap from="$(arg robot_name)_kobuki_safety_controller/cmd_vel" to="cmd_vel_mux/input/safety_controller"/>
    <remap from="$(arg robot_name)_kobuki_safety_controller/events/bumper" to="events/bumper"/>
    <remap from="$(arg robot_name)_kobuki_safety_controller/events/cliff" to="events/cliff"/>
    <remap from="$(arg robot_name)_kobuki_safety_controller/events/wheel_drop" to="events/wheel_drop"/>
  </node>

  <!-- robot_pose_ekf -->
 <node  pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf" >
    <param name="freq" value="30.0"/>
    <param name="sensor_timeout" value="1.0"/>
    <param name="publish_tf" value="true"/>
    <param name="odom_used" value="true"/>
    <param name="imu_used" value="true"/>
    <param name="vo_used" value="false"/>
    <param name="output_frame" value="odom"/>
    <remap from="odom" to="$(arg robot_name)/odom"/>
  </node>

  <!-- Localization -->
  <arg name="initial_pose_x" default="$(arg x)"/>
  <arg name="initial_pose_y" default="$(arg y)"/>
  <arg name="initial_pose_a" default="0.0"/>

  <include file="$(find lucia_sim_2014)/launch/amcl/amcl.launch.xml">
    <arg name="initial_pose_x" value="$(arg initial_pose_x)"/>
    <arg name="initial_pose_y" value="$(arg initial_pose_y)"/>
    <arg name="initial_pose_a" value="$(arg initial_pose_a)"/>
    <arg name="robot_name"  value="$(arg robot_name)" />
  </include>

  <!-- move_base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" >
    <remap from="map" to="/map" />
    <rosparam file="$(find lucia_sim_2014)/launch/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find lucia_sim_2014)/launch/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find lucia_sim_2014)/launch/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find lucia_sim_2014)/launch/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find lucia_sim_2014)/launch/param/base_local_planner_params.yaml" command="load" />
    <rosparam file="$(find lucia_sim_2014)/launch/param/move_base_params.yaml" command="load" />
    <remap from="cmd_vel" to="$(arg robot_name)/commands/velocity"/>
    <remap from="map" to="/map" />
  </node>

  <!-- lucia services -->
  <node pkg="lucia_sim_2014" type="lucia_services" name="lucia_services" output="screen">
    <remap from="commands/velocity" to="$(arg robot_name)/commands/velocity"/>
    <param name="robot_id" type="string" value="$(arg robot_name)" />
   </node>

</launch>























